### PR TITLE
kubeadm: remove duplicate DNS names and IP addresses from generated certificates

### LINKED ISSUE #92751: Duplicate altnames in certs which generated by kubeadm
**What happened**:

Duplicate altnames in certs which generated by kubeadm
image

**What you expected to happen**:

No Duplicate altnames in certs which generated by kubeadm

**How to reproduce it (as minimally and precisely as possible)**:

Anytime

**Anything else we need to know?**:

image

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by QianChenglong**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

fix duplicate altnames in cert

**Which issue(s) this PR fixes**:

Fixes #92751 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

**Release note:**
```release-note
kubeadm: remove duplicate DNS names and IP addresses from generated certificates
```

### CODE REVIEW

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 1]
```diff
@@ -427,8 +431,13 @@ func getAltNames(cfg *kubeadmapi.InitConfiguration, certName string) (*certutil.
 
 	// create AltNames with defaults DNSNames/IPs
 	altNames := &certutil.AltNames{
-		DNSNames: []string{cfg.NodeRegistration.Name, "localhost"},
```
Comment (SataQiu) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:430]: "localhost" lost?

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 2]
```diff
@@ -53,6 +53,33 @@ type AltNames struct {
 	IPs      []net.IP
 }
 
+// CleanAltNames cleans duplicate items in DNSNames and IPs.
+func CleanAltNames(altNames *AltNames) {
+	if altNames == nil {
+		return
+	}
+
+	dnsNamesKeys := make(map[string]struct{})
```
Comment (SataQiu) [inline staging/src/k8s.io/client-go/util/cert/cert.go:62]: FYI: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets
There are some utils you can consider to use.
```go
sets.NewString(altNames.DNSNames...).List()
```

File: cmd/kubeadm/app/phases/certs/certlist.go [hunk 3]
```diff
@@ -21,7 +21,6 @@ import (
 	"crypto/x509"
 
 	"github.com/pkg/errors"
-
```
Comment (SataQiu) [inline cmd/kubeadm/app/phases/certs/certlist.go:24]: Why delete the blank line? We prefer to use a blank line to distinguish different import blocks.

**Review by SataQiu — COMMENTED** [review]
Thanks @QianChenglong !
Add some comments.

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 4]
```diff
@@ -383,6 +381,12 @@ func GetAPIServerAltNames(cfg *kubeadmapi.InitConfiguration) (*certutil.AltNames
 			advertiseAddress,
 		},
 	}
+	// cfg.NodeRegistration.Name may use ip!
```
Comment (neolit123) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:384]: thank you for the PR,
i don't think i agree with this change and the similar one in getAltNames() . the comment that "node names can be IPs" is wrong.

the Node name is supposed to be a string "that can be used as a DNS subdomain name as defined in RFC 1123"
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

https://github.com/kubernetes/kubernetes/blob/47a9952337dada2cd57f3b2b3aeccac500277318/cmd/kubeadm/app/apis/kubeadm/validation/validation.go#L107

however, this function is not perfectly following RFC 1123 and it allows the strings to start with a number, which is wrong.

the node name *should be* a hostname, not an IP address and we should not parse it as an IP address.
if you want to add the same IP address in altNames you should do so explicitly using the extraSANs kubeadm feature.


/hold

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 5]
```diff
@@ -427,8 +431,13 @@ func getAltNames(cfg *kubeadmapi.InitConfiguration, certName string) (*certutil.
 
 	// create AltNames with defaults DNSNames/IPs
 	altNames := &certutil.AltNames{
-		DNSNames: []string{cfg.NodeRegistration.Name, "localhost"},
-		IPs:      []net.IP{advertiseAddress, net.IPv4(127, 0, 0, 1), net.IPv6loopback},
```
Comment (neolit123) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:431]: losing the above two lines is a regression.

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 6]
```diff
@@ -53,6 +53,33 @@ type AltNames struct {
 	IPs      []net.IP
 }
 
+// CleanAltNames cleans duplicate items in DNSNames and IPs.
+func CleanAltNames(altNames *AltNames) {
```
Comment (neolit123) [inline staging/src/k8s.io/client-go/util/cert/cert.go:57]: in general i like the idea of cleaning up duplicate entries in the altNames.
but please name this function `RemoveDuplicateAltNames`.

also kubeadm supports wildcards such as `127.0.*.*` do we care to remove `127.0.0.1` if `127.0.*.*` is in the list already?
possibly not.

**Review by neolit123 — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go (hunk 1 above)
Comment (QianChenglong) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Thanks for pointing out the error, leave it as it is.

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 2 above)
Comment (QianChenglong) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: ok

File: cmd/kubeadm/app/phases/certs/certlist.go (hunk 3 above)
Comment (QianChenglong) [reply cmd/kubeadm/app/phases/certs/certlist.go]: Sorry for bringing our programming style, I would like to ask what tool is used here to automatically control imports？

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go (hunk 4 above)
Comment (QianChenglong) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Thanks for pointing out the error, leave it as it is.

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 6 above)
Comment (QianChenglong) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: RemoveDuplicateAltNames is indeed more clear.
For the time being, only the repetition of the character form is removed, and logical repetition may be necessary under observation.

**Comment by QianChenglong** [comment]
Thank you for all comments, it has been revised, please confirm.

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 7]
```diff
@@ -53,6 +54,27 @@ type AltNames struct {
 	IPs      []net.IP
 }
 
+// RemoveDuplicateAltNames removes duplicate items in altNames.
+func RemoveDuplicateAltNames(altNames *AltNames) {
```
Comment (neolit123) [inline staging/src/k8s.io/client-go/util/cert/cert.go:58]: i think i saw a unit test for the old function? we should still have one (e.g. called TestRemoveDuplicateAltNames)

Comment (QianChenglong) [reply staging/src/k8s.io/client-go/util/cert/cert.go:58]: Sorry for missing that.

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 8]
```diff
@@ -35,7 +35,6 @@ import (
 	"time"
 
 	"github.com/pkg/errors"
-
```
Comment (neolit123) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:38]: one minor nit: we should not remove this whitespace line.

Comment (QianChenglong) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Done

**Comment by neolit123** [comment]
thanks
/ok-to-test

**Comment by neolit123** [comment]
please have a look at the failures:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92753/pull-kubernetes-verify/1278877246051848192
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92753/pull-kubernetes-bazel-test/1278877245212987395

you might have to call ./hack/update-bazel.sh and commit the changes.
the other failure is about the boilerplate:

> Boilerplate header is wrong for: /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/util/cert/cert_test.go

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 7 above)
Comment (neolit123) [inline staging/src/k8s.io/client-go/util/cert/cert.go:58]: please move this function and the unit tests to:

cmd/kubeadm/app/util/pkiutil/pki_helpers.go
cmd/kubeadm/app/util/pkiutil/pki_helpers_test.go

let's keep it to the kubeadm use case for now, and later if it's someone needs it we can move it to client-go.

Comment (QianChenglong) [reply staging/src/k8s.io/client-go/util/cert/cert.go:58]: ok

**Comment by neolit123** [comment]
this looks fine to me.
/retest

do we have comments by others? cc @kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by neolit123** [comment]
@QianChenglong please instead of `NONE` in the PR description add:

> kubeadm: remove duplicate DNS names and IP addresses from generated certificates

**Comment by QianChenglong** [comment]
> @QianChenglong please instead of `NONE` in the PR description add:
> 
> > kubeadm: remove duplicate DNS names and IP addresses from generated certificates

ok

**Review by dixudx — APPROVED** [review]
/lgtm

**Comment by neolit123** [comment]
/retitle kubeadm: remove duplicate DNS names and IP addresses from generated certificates

**Pull request merged**